File: binary_search/koko_eating_bananas.java
class Solution
{
    int limit;

    public boolean check(double k, int[] piles)
    {
        int hours = 0;

        for (double banana : piles)
        {
            hours += Math.ceil(bananas / k);
        }

        return hours <= limit;
    }

    public int minEatingSpeed(int[] piles, int h)
    {
        limit = h;

        int left = 1;
        int right = 0;

        for (int bananas : piles)
        {
            right = Math.max(right, bananas);
        }

        while (left <= right)
        {
            int mid = left + (right - left) / 2;

            if (check(mid, piles))
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }

        return left;
    }
}

File: binary_search/koko_eating_bananas.py
import math


def minEatingSpeed(piles: list[int], h: int) -> int:
    def check(k):
        hours = 0
        for bananas in piles:
            hours += math.ceil(bananas / k)
        
        return hours <= h
    
    left = 1
    right = max(piles)

    while left <= right:
        mid = (left + right) // 2
        
        if check(mid):
            right = mid - 1
        else:
            left = mid + 1
    
    return left